[MM-52828] Migrate "components/admin_console/team_channel_settings/group/index.js" to typescript

ridwankabeer435 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/23432
https://mattermost.atlassian.net/browse/MM-52828


#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note

```

mattermost-build commented:
Hello @ridwankabeer435,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

ridwankabeer435 commented:
/check-cla

sinansonmez commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
```diff
@@ -0,0 +1,46 @@
+import {GlobalState} from 'types/store';
+
+type Actions = {
+
```
nit: no extra lines please

sinansonmez commented on webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx:
```diff
@@ -62,7 +62,7 @@ export const ChannelGroups: React.FunctionComponent<ChannelGroupsProps> = (props
                 <GroupList
                     channel={channel}
-                    groups={groups}
+                    groups={groups as Group[]}
```
Why do we need a type casting here?

ridwankabeer435 replied:
I kept getting the following error without the type casting:

`Type 'Partial<Group>[]' is not assignable to type 'Group[]'.
  Type 'Partial<Group>' is not assignable to type 'Group'.
    Types of property 'id' are incompatible.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.ts(2322)`

sinansonmez replied:
It seems to me id is creating an issue because it can be also undefined. GroupList doesnt like that. We can make sure that id is not undefined by assigning it to empty string when it is undefined. Alternatively, we can change the type of groups as `Partial<Group>[]` in GroupList.

ridwankabeer435 replied:
@sinansonmez I tried changing the type of groups to `Partial<Group>[]` but then `AbstrasctList` inside `GroupList` does not like that because of its `data: Array<Group | TeamWithMembership>`. I then tried changing that to `data: Array<Partial<Group> | Partial<TeamWithMembership>>`, but then TypeScript's transpiler complains at `return this.props.data.slice(offset, offset + PAGE_SIZE).map(this.props.renderRow);` inside the renderRows functions, which expects a type of `JSX.Element | JSX.Element[]`

This is the error I get:
`Argument of type '(item: Group | TeamWithMembership) => Element' is not assignable to parameter of type '(value: Partial<Group> | Partial<TeamWithMembership>, index: number, array: (Partial<Group> | Partial<TeamWithMembership>)[]) => Element'.
  Types of parameters 'item' and 'value' are incompatible.
    Type 'Partial<Group> | Partial<TeamWithMembership>' is not assignable to type 'Group | TeamWithMembership'.
      Type 'Partial<Group>' is not assignable to type 'Group | TeamWithMembership'.
        Type 'Partial<Group>' is not assignable to type 'Group'.
          Types of property 'id' are incompatible.
            Type 'string | undefined' is not assignable to type 'string'.
              Type 'undefined' is not assignable to type 'string'.`

ridwankabeer435 replied:
I have updated `channel_groups.tsx` to set default values for each missing `Partial<Group>` attribute before passing the `groups` prop to `GroupList`. However, I still remain uncertain about the default values. Would you suggest making each attribute in the `Group` type optional?

sinansonmez replied:
yes, we can keep the default values in the test

sinansonmez commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
```diff
@@ -0,0 +1,46 @@
+    return {
+        actions: bindActionCreators<ActionCreatorsMapObject<ActionFunc>, Actions>({
+            getData: () => getGroups(),
+
```
nit: no extra line please

sinansonmez commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
```diff
@@ -0,0 +1,46 @@
+function mapDispatchToProps(dispatch: Dispatch) {
+    return {
+        actions: bindActionCreators<ActionCreatorsMapObject<ActionFunc>, Actions>({
+            getData: () => getGroups(),
```
`getGroups` was not used in the older version. Why do we need to use it here or am I missing something?

ridwankabeer435 replied:
I was getting compilation errors with the older version of the function. I guessed that using  `getGroups` action would give us the same result as `Promise.resolve()`. Do you still want to retain the older version of this function?

sinansonmez replied:
I dont think `getGroups` is the correct action to be used here. Because if you check `GroupList` action is passed down to `AbstractList` component and `getGroup` type definition is different there. 

In the meantime I am also wondering how this component is working properly with  `Promise.resolve()`. 

Therefore I would request another review from the internal member who knows this area of the code base cc: @M-ZubairAhmed

sinansonmez replied:
@M-ZubairAhmed any idea here?

ridwankabeer435 replied:
@M-ZubairAhmed, would really appreciate your insights about the most suitable action for getting groups. This is my first time contributing to the Mattermost codebase

M-ZubairAhmed replied:
Oops sorry i dont know how i missed this notification for so long. Let me answer you tomorrow

M-ZubairAhmed replied:
lets use the same promise.resolve, i think the component should have been properly added the fallback when getData was not defined.

sinansonmez reviewed (changes requested):
Thanks for the effort. I just have a few small comments

M-ZubairAhmed commented:
/release-note-none

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
@ridwankabeer435 can you please update your fork to `mattermost`

M-ZubairAhmed commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
Lets follow `webapp/channels/src/components/team_controller/index.ts` to implement PropsFromRedux approach here

ridwankabeer435 replied:
I think the only major change would be to add `actions: {
            getData: () => Promise.resolve(ownProps.groups),
        },` under `mapStateToProps()` before passing the action to `PropsFromRedux`

ridwankabeer435 commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
```diff
@@ -32,13 +26,12 @@ function mapStateToProps(state: GlobalState, ownProps: OwnProps) {
 
-export default connect(mapStateToProps, mapDispatchToProps)(List);
+export type PropsFromRedux = ConnectedProps<typeof connector>;
 
```
@M-ZubairAhmed is this what you had in mind? I am getting the following error in `channel_groups.tsx` and `team_groups.tsx` after making the changes: `Property 'actions' is missing in type '{ team: Team; isModeSync: boolean; groups: Group[]; totalGroups: number; onGroupRemoved: (groupId: string) => void; setNewGroupRole: (groupId: string) => void; type: string; isDisabled: boolean | undefined; }' but required in type 'Omit<ClassAttributes<GroupList> & Props, "data" | "total" | "emptyListTextId" | "emptyListTextDefaultMessage" | "removeGroup" | "setNewGroupRole">'`;

basically `GroupList` is now expecting its parent components to pass an implementation of `getData: () => Promise<Group[]>` as a prop value to `GroupList` 's `actions` prop.

ridwankabeer435 replied:
Please do not run any integration and end-to-end test at this point.

M-ZubairAhmed replied:
@ridwankabeer435 i have pulled this branch locally and here are my observations/suggestions

- As seen in https://github.com/mattermost/mattermost/pull/23511#discussion_r1245502729 Lets export this ownProps as well.
- In https://github.com/mattermost/mattermost/pull/23511#discussion_r1245503014 so essentially you Props will just be a combinations of OwnProps and PropsFromRedux
- You might need to add more ownProp types in index file
- Cross check the function return signature of mapDispatchToProps https://github.com/mattermost/mattermost/pull/23511#discussion_r1245504798 the new one should be the same
- Because this is a weird case lets hard type cast the `getData: () => Promise.resolve() as any,` 

I wrote everything here to condense the message clearly in one place

M-ZubairAhmed commented on webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:
```diff
@@ -40,7 +40,7 @@ const Header = () => {
 };
 
-interface Props {
+type Props = {
```
[2]

M-ZubairAhmed commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.js:
```diff
@@ -1,30 +0,0 @@
-    };
-}
-
-function mapDispatchToProps() {
-    return {
-        actions: {
-            getData: () => Promise.resolve(),
-        },
-    };
-}
```
[3]

ridwankabeer435 replied:
I am not sure what you meant by the [n] comments. Are those potential problem points within the code commits that may be requiring team_groups and channel_groups to pass an action that returns Promise<Group[]>? I am really lost at this point :(

M-ZubairAhmed replied:
sorry i just used this a reference points to explain in https://github.com/mattermost/mattermost/pull/23511#discussion_r1245506853
apologies for causing this as confusion. But you are doing really great job 💯

ridwankabeer435 commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
```diff
@@ -18,16 +20,20 @@ type OwnProps = {
         setNewGroupRole: ownProps.setNewGroupRole,
+        actions: {
+            getData: () => Promise.resolve(ownProps.groups),
+        },
```
Passing `ownProps.groups` to `Promise.resolve` for `getData()` action  pretty much solved most of the issues

ridwankabeer435 commented on webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:
```diff
@@ -18,16 +20,20 @@ type OwnProps = {
 
 const mapDispatchToProps = {
-    getData: () => Promise.resolve(),
+    getData: () => Promise.resolve() as any,
```
not sure if we need to cast `Promise.resolve()` as any at this point. But I suppose it can work as insurance.

M-ZubairAhmed replied:
yes this is an exceptional case where even I have no idea why we had this in the first place. But fixing this in this pr is out of scope

M-ZubairAhmed commented on webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:
```diff
@@ -40,18 +40,18 @@ const Header = () => {
 };
 
-interface Props {
+type Props = {
```
this can be just Props = OwnProps & PropsFromRedux

ridwankabeer435 replied:
Hi @M-ZubairAhmed, I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running `npm test` on the test file): `Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(GroupList) in connect options.`

ridwankabeer435 commented on webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:
```diff
@@ -47,7 +47,7 @@ type Props = {
+        getData: () => Promise<any>;
     };
     removeGroup: (gid: string) => void;
     setNewGroupRole: (gid: string) => void;
```
kept most of the original props as `group_list_test` starts complaining if I use `type Props = OwbnProps & PropsFromRedux` instead. Not sure if we want to change `group_list_test` to accommodate using `type Props = OwbnProps & PropsFromRedux`.

M-ZubairAhmed replied:
I would fix type issue in `group_list_test` instead

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

M-ZubairAhmed commented:
@ridwankabeer435 do you need help in getting this to finish line?

ridwankabeer435 commented:
> 
I'd like to give it one last shot to fix the issue first. Apologies for leaving the PR hanging for so long

M-ZubairAhmed commented:
@ridwankabeer435 anything i can do to help you out

ridwankabeer435 commented:
> @ridwankabeer435 do you need help in getting this to finish line?

I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(GroupList) in connect options.

I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

M-ZubairAhmed commented:
> > @ridwankabeer435 do you need help in getting this to finish line?
> 
> I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> 
> I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests

Can you push the code so we can verify whats the issue there

ridwankabeer435 commented:
> > > @ridwankabeer435 do you need help in getting this to finish line?
> > 
> > 
> > I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> > I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests
> 
> Can you push the code so we can verify whats the issue there

I have pushed whatever changes I have made so far. I sincerely apologize for my lack of commitment over the past few month

M-ZubairAhmed commented:
> > > > @ridwankabeer435 do you need help in getting this to finish line?
> > > 
> > > 
> > > I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> > > I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests
> > 
> > 
> > Can you push the code so we can verify whats the issue there
> 
> I have pushed whatever changes I have made so far. I sincerely apologize for my lack of commitment over the past few month

Thank you let me have a look but before that can you please the conflicts for the CI to run?

ridwankabeer435 commented:
> > > > > @ridwankabeer435 do you need help in getting this to finish line?
> > > > 
> > > > 
> > > > I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> > > > I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests
> > > 
> > > 
> > > Can you push the code so we can verify whats the issue there
> > 
> > 
> > I have pushed whatever changes I have made so far. I sincerely apologize for my lack of commitment over the past few month
> 
> Thank you let me have a look but before that can you please the conflicts for the CI to run?

I have resolved the conflicts and styling issues

M-ZubairAhmed commented:
/update-branch
